### PR TITLE
Update cluster selection and reconcile logic

### PR DESCRIPTION
**Pull request by irinamihai**
Description:
- update cluster selection to only include in the remediationPlan those clusters that are present in the CR and NonCompliant in the at least one of the specified managed policies
- have just one copy of each managed policy to use for the upgrade; for each of those copied policies, have one placement rule and one placement binding. Before the upgrade starts, each placement rule matches 0 clusters and the copied inform policies are set to enforce.
- in the main reconcile loop, once the upgrade starts, for each cluster in a batch, start applying one policy at a time (starting with the first policy) by adding the cluster to the corresponding placement rule.
- add new member to UpgradeStatus: RemediationPlanForBatch that holds the policy index for each cluster in the batch
- once a batch is done, the placement rules are cleared (go back to matching no cluster)
- TODOs have been added to places that need either cleanup or further restructuring

### CODE REVIEW

File: controllers/clustergroupupgrade_controller.go [hunk 1]
```diff
@@ -439,19 +583,19 @@ func (r *ClusterGroupUpgradeReconciler) ensureBatchPlacementRules(ctx context.Co
...
 	u.Object = map[string]interface{}{
 		"metadata": map[string]interface{}{
-			"name":      clusterGroupUpgrade.Name + "-batch" + strconv.Itoa(batchIndex),
+			"name":      clusterGroupUpgrade.Name + "-" + managedPolicy.GetName(),
 			"namespace": clusterGroupUpgrade.Namespace,
 			"labels": map[string]interface{}{
 				"app": "openshift-cluster-group-upgrades",
 				"openshift-cluster-group-upgrades/clusterGroupUpgrade": clusterGroupUpgrade.Name,
-				"openshift-cluster-group-upgrades/batch":               strconv.Itoa(batchIndex),
+				"openshift-cluster-group-upgrades/forPolicy":           clusterGroupUpgrade.Name + "-" + managedPolicy.GetName(),
```
Comment (imiller0) [inline controllers/clustergroupupgrade_controller.go:598]: Do we need to include the namespace here as well? When managing multiple clusters we could end up with site-1/policy-A and site-2/policy-A (ie same policy name under 2 different namespaces). This may apply to the placementRule and placementBinding names as well. We need to ensure that there is not a name collision when managing the same named policy in two different cluster namespaces.

File: controllers/clustergroupupgrade_controller.go [hunk 2]
```diff
@@ -439,19 +583,19 @@ func (r *ClusterGroupUpgradeReconciler) ensureBatchPlacementRules(ctx context.Co
...
+	return &placementRule, nil
 }
 
-func (r *ClusterGroupUpgradeReconciler) newBatchPlacementRule(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, batch []string, batchIndex int) (*unstructured.Unstructured, error) {
+func (r *ClusterGroupUpgradeReconciler) newBatchPlacementRule(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, managedPolicy *unstructured.Unstructured) (*unstructured.Unstructured, error) {
 	u := &unstructured.Unstructured{}
 	u.Object = map[string]interface{}{
 		"metadata": map[string]interface{}{
-			"name":      clusterGroupUpgrade.Name + "-batch" + strconv.Itoa(batchIndex),
+			"name":      clusterGroupUpgrade.Name + "-" + managedPolicy.GetName(),
```
Comment (imiller0) [inline controllers/clustergroupupgrade_controller.go:593]: Same name calculation here and below. Maybe create the name in a var and use it both places?

File: controllers/clustergroupupgrade_controller.go [hunk 3]
```diff
@@ -662,6 +802,25 @@ func (r *ClusterGroupUpgradeReconciler) newBatchOperatorUpgradePolicy(ctx contex
...
+func (r *ClusterGroupUpgradeReconciler) isPolicyCompliant(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) (bool, error) {
+	var result bool
+	policy, err := r.getBatchPlatformUpgradePolicy(ctx, clusterGroupUpgrade)
+	if err != nil {
+		return false, err
+	}
+
+	if policy != nil {
+		statusObject := policy.Object["status"].(map[string]interface{})
+		if statusObject["compliant"] == nil || statusObject["compliant"] == "NonCompliant" {
```
Comment (imiller0) [inline controllers/clustergroupupgrade_controller.go:814]: Use a constant for "NonCompliant"

File: controllers/clustergroupupgrade_controller.go [hunk 4]
```diff
@@ -717,15 +876,16 @@ func (r *ClusterGroupUpgradeReconciler) areAllPoliciesCompliant(ctx context.Cont
...
-func (r *ClusterGroupUpgradeReconciler) ensureBatchPlacementBindings(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, batchPlacementRules []string, batchPolicies []string, batchIndex int) ([]string, error) {
-	var placementBindings []string
+func (r *ClusterGroupUpgradeReconciler) ensureBatchPlacementBinding(
+	ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, batchPlacementRule string, managedPolicy string) (*string, error) {
+	var placementBinding string
 
-	// ensure batch placement bindings
-	pb, err := r.newBatchPlacementBinding(ctx, clusterGroupUpgrade, batchPlacementRules, batchPolicies, batchIndex)
+	// Ensure batch placement bindings.
+	pb, err := r.newBatchPlacementBinding(ctx, clusterGroupUpgrade, batchPlacementRule, managedPolicy)
```
Comment (imiller0) [inline controllers/clustergroupupgrade_controller.go:884]: If "pb" is only used in one branch of conditional, move this logic inside that branch.

File: controllers/clustergroupupgrade_controller.go [hunk 5]
```diff
@@ -973,34 +1141,148 @@ func (r *ClusterGroupUpgradeReconciler) reconcileResources(ctx context.Context,
...
 
-func (r *ClusterGroupUpgradeReconciler) buildRemediationPlan(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) {
+func (r *ClusterGroupUpgradeReconciler) isClusterNonCompliantWithPolicy(clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, clusterName string, policy *unstructured.Unstructured) bool {
+	// Get the status part of the policy.
+	statusObject := policy.Object["status"].(map[string]interface{})
+	// Get the part of the status that holds the clusters' compliance.
+	statusCompliance := statusObject["status"]
+
+	// If all the clusters are compliant, return nil.
+	if statusCompliance == nil {
```
Comment (imiller0) [inline controllers/clustergroupupgrade_controller.go:1151]: Is a nil status really equivalent to the policy being compliant?

File: controllers/clustergroupupgrade_controller.go [hunk 6]
```diff
@@ -973,34 +1141,148 @@ func (r *ClusterGroupUpgradeReconciler) reconcileResources(ctx context.Context,
...
+		return false
+	}
+
+	subStatus := statusCompliance.([]interface{})
+
+	// Loop through all the NonCompliant clusters in the policy's status.
+	for _, crtSubStatusCrt := range subStatus {
+		crtSubStatusMap := crtSubStatusCrt.(map[string]interface{})
+		// If the cluster is NonCompliant, add it to the list of NonCompliant clusters.
```
Comment (imiller0) [inline controllers/clustergroupupgrade_controller.go:1160]: The comment says it should be added to the list but that is not done here.

File: controllers/clustergroupupgrade_controller.go [hunk 7]
```diff
@@ -973,34 +1141,148 @@ func (r *ClusterGroupUpgradeReconciler) reconcileResources(ctx context.Context,
...
+    placementRule: operator-placementrules
+  status: ([]interface{})
+  - clustername: spoke3
+    clusternamespace: spoke3
+    compliant: NonCompliant
+  - clustername: spoke4
+    clusternamespace: spoke4
+    compliant: NonCompliant
+*/
+func (r *ClusterGroupUpgradeReconciler) getClustersNonCompliantWithPolicy(policy *unstructured.Unstructured) []string {
```
Comment (imiller0) [inline controllers/clustergroupupgrade_controller.go:1185]: I wonder if ACM has any golang libraries that we could leverage/use for some of these functions?

**Review by imiller0 — COMMENTED** [review]
Nice work!
Please add unit test coverage as well.
Thanks!

File: api/v1alpha1/clustergroupupgrade_types.go [hunk 8]
```diff
@@ -80,6 +77,7 @@ type UpgradeStatus struct {
 	CompletedAt             metav1.Time    `json:"completedAt,omitempty"`
 	CurrentBatch            int            `json:"currentBatch,omitempty"`
 	CurrentBatchStartedAt   metav1.Time    `json:"currentBatchStartedAt,omitempty"`
+	RemediationPlanForBatch map[string]int `json:"remediationPlanForBatch,omitempty"`
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply api/v1alpha1/clustergroupupgrade_types.go:80]: For sure, will do.

File: controllers/clustergroupupgrade_controller.go [hunk 9]
```diff
@@ -281,24 +266,189 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
 	return nextReconcile, nil
 }
 
+func (r *ClusterGroupUpgradeReconciler) initializeRemediationPolicyForBatch(
+	clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) {
+
+	if clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt.IsZero() {
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:272]: I agree, will move it.

File: controllers/clustergroupupgrade_controller.go [hunk 10]
```diff
@@ -281,24 +266,189 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
...
+			// Maybe continue to next cluster in batch or just move to the next policy???
+			return err, false
+		}
+
+		// Check if current cluster is compliant or not for its current policy.
+		isBatchClusterNonCompliantForPolicy := r.isClusterNonCompliantWithPolicy(clusterGroupUpgrade, batchClusterName, currentManagedPolicy)
+
+		r.Log.Info("isBatchClusterNonCompliantForPolicy", "isBatchClusterNonCompliantForPolicy", isBatchClusterNonCompliantForPolicy)
+		// If the cluster is compliant for the policy, move to the next policy index.
+		if isBatchClusterNonCompliantForPolicy == false {
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:314]: Ugh, I also thought about that, really. I'll try and change it so that the code is easier to read.

File: controllers/clustergroupupgrade_controller.go [hunk 11]
```diff
@@ -281,24 +266,189 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
...
+
+		// Set the time for when the batch started updating.
+		clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt = metav1.Now()
+
+		r.Log.Info(">>>> InitializeRemediationPolicyForBatch ", "remediationPlanForBatch", clusterGroupUpgrade.Status.Status.RemediationPlanForBatch)
+		r.Log.Info("NEW CurrentBatchStartedAt", "clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt", clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt)
+	}
+}
+
+func (r *ClusterGroupUpgradeReconciler) getNextRemediationPoliciesForBatch(
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:289]: Yes, I like that. Will add. I have that for another function below. It would be nice to have something like this for every function. Will add it as part of the cleanup.

File: controllers/clustergroupupgrade_controller.go [hunk 12]
```diff
@@ -147,61 +153,35 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
...
 
-				isBatchPlatformUpgradeComplete, err = r.isBatchPlatformUpgradePolicyCompliant(ctx, clusterGroupUpgrade)
-				if err != nil {
-					return ctrl.Result{}, err
-				}
-			}
-
-			operatorUpgradePolicy, err := r.getBatchOperatorUpgradePolicy(ctx, clusterGroupUpgrade)
+			// Add the needed cluster names to upgrade to the appropriate placement rule.
+			err = r.addClustersToPlacementRule(ctx, clusterGroupUpgrade)
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:168]: I see your point. My approach was to remove all clusters from the placementRules once the batch is done upgrading. The reason I like this approach is because technically, when a batch is done upgrading, all the copied policies are compliant, if we ever want to check. I think it would be nice to have this option to check in the future. What do you think?

Is it ok if I document the current functionality in a method level comment, as we discussed in the comments right above?

File: controllers/clustergroupupgrade_controller.go [hunk 13]
```diff
@@ -147,61 +153,35 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
...
-
-			}
 
 			if isBatchComplete {
-				r.Log.Info("Batch upgrade completed")
+				r.Log.Info("Upgrade completed for batch", "batchIndex", clusterGroupUpgrade.Status.Status.CurrentBatch)
 				if clusterGroupUpgrade.Status.Status.CurrentBatch < len(clusterGroupUpgrade.Status.Policies) {
 					clusterGroupUpgrade.Status.Status.CurrentBatch++
+					clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt = metav1.Time{}
+					r.cleanupPlacementRules(ctx, clusterGroupUpgrade)
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:178]: Yes, I agree.

File: controllers/clustergroupupgrade_controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:272]: I strongly agree, I don't really like seeing these through the code either, it impacts the readability. I will add them to the cleanup as they are not something to block functionality, do you agree?

File: controllers/clustergroupupgrade_controller.go [hunk 14]
```diff
@@ -281,24 +266,189 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
...
 }
 
+func (r *ClusterGroupUpgradeReconciler) initializeRemediationPolicyForBatch(
+	clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) {
+
+	if clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt.IsZero() {
+		clusterGroupUpgrade.Status.Status.RemediationPlanForBatch = make(map[string]int)
+		batchIndex := clusterGroupUpgrade.Status.Status.CurrentBatch - 1
+
+		// By default, assume all clusters start applying all policies from the first policy.
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:276]: Yes, I had this approach initially, where I started with the first policy, but we all discussed (sorry, you were away on vacation) that it takes quite a lot of computing to determine which is the first managed policy in which the cluster is non compliant (we need to go through all the policies and check all the NonCompliant clusters...), so we decided to just go through all the policies from the start.

File: controllers/clustergroupupgrade_controller.go [hunk 15]
```diff
@@ -281,24 +266,189 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
...
+
+	for _, batchClusterName := range clusterGroupUpgrade.Status.RemediationPlan[batchIndex] {
+		// Get the index and name of the current policy being applied for the current cluster in the batch.
+		currentPolicyIndex := clusterGroupUpgrade.Status.Status.RemediationPlanForBatch[batchClusterName]
+		currentManagedPolicyName := clusterGroupUpgrade.Spec.ManagedPolicies[currentPolicyIndex]
+		r.Log.Info("Current policy name for cluster ", "cluster name", batchClusterName, "policyName", currentManagedPolicyName)
+
+		err, currentManagedPolicy := r.getPolicyByName(ctx, currentManagedPolicyName, clusterGroupUpgrade.Namespace)
+		if err != nil {
+			// Maybe continue to next cluster in batch or just move to the next policy???
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:305]: In here it would only be an error with getting a policy, that's why I was a bit unsure. There are many other places where we return errors in case we cannot get a managedCluster, a placement rule/binding. I think it would be worth having a better handling of these errors, propagating the main error to the main reconcile loop and include it in the status somehow.
I propose we add this to the cleanup as well.

File: controllers/clustergroupupgrade_controller.go [hunk 16]
```diff
@@ -281,24 +266,189 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
...
+	for _, plr := range placementRules.Items {
+		r.Log.Info("Cleanup placementRule", "plrName", plr.GetName())
+		placementRuleSpecClusters := plr.Object["spec"].(map[string]interface{})
+		placementRuleSpecClusters["clusters"] = nil
+		placementRuleSpecClusters["clusterReplicas"] = 0
+
+		err = r.Client.Update(ctx, &plr)
+		if err != nil {
+			r.Log.Info("Error cleaning up placementRule", "err", err)
+			return err
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:421]: Yes, good idea. Will do.

File: controllers/clustergroupupgrade_controller.go [hunk 17]
```diff
@@ -399,14 +543,14 @@ func (r *ClusterGroupUpgradeReconciler) createNewPolicyFromStructure(
 	return nil
 }
 
-func (r *ClusterGroupUpgradeReconciler) ensureBatchPlacementRules(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, batch []string, batchIndex int) ([]string, error) {
-	var placementRules []string
+func (r *ClusterGroupUpgradeReconciler) ensureBatchPlacementRule(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, managedPolicy *unstructured.Unstructured) (*string, error) {
+	var placementRule string
 
-	pr, err := r.newBatchPlacementRule(ctx, clusterGroupUpgrade, batch, batchIndex)
+	pr, err := r.newBatchPlacementRule(ctx, clusterGroupUpgrade, managedPolicy)
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:549]: It is used a bit below, with pr.GetName(), but the name can be obtained from the managedPolicy anyway, so I can move it as you suggested. Thanks!

File: controllers/clustergroupupgrade_controller.go (hunk 2 above)
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:593]: Yes, I also thought about that. Will do.

File: controllers/clustergroupupgrade_controller.go (hunk 1 above)
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:598]: Hmm, we have the namespace on line 594 above. Isn't that not enough? I think I included the Namespace when creating/getting all the resources, except for the managedClusters. I will double check.

File: controllers/clustergroupupgrade_controller.go (hunk 5 above)
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:1151]: To be honest, I am not 100% sure. What I relied upon was the status.status of this policy below which is nil if no clusters are NonCompliant with it (again with the reverse logic here, sorry).
```
status:
  compliant: NonCompliant
  placement:
  - placementBinding: binding-policy1
    placementRule: placement-policy1
  status:
  - clustername: spoke1
    clusternamespace: spoke1
    compliant: NonCompliant
  - clustername: spoke6
    clusternamespace: spoke6
    compliant: NonCompliant
```

I suppose that instead I could check for status.compliant, but my worry is that if somehow there is no placementRule/placementBinding for the policy, then we don't have access to that field either, that's why I chose this option.

File: controllers/clustergroupupgrade_controller.go (hunk 6 above)
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:1160]: Oops, missed it when I restructured this function. Thank you.

File: controllers/clustergroupupgrade_controller.go (hunk 7 above)
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:1185]: Yes, it does. I actually tried playing with one on Friday, but ended up getting some errors and decided to better work on having something functional first and then look more into it. What I found was a library for placementRules.

Check this out: https://github.com/RamenDR/ramen/blob/aaab22aea138286a671046675f2526f405f25986/controllers/drplacementcontrol_controller.go#L977, https://github.com/RamenDR/ramen/blob/aaab22aea138286a671046675f2526f405f25986/controllers/drplacementcontrol_controller.go#L882
using this import: github.com/open-cluster-management/multicloud-operators-placementrule/pkg/apis/apps/v1

Unfortunately I didn't save the exact error, but it was complaining about the specific placementRule kind used by that package. I suspect an extra change was needed in SetupWithManager, but I chose to move forward with my stuff after spending some time and not being able to figure it out :disappointed: It's on my list of improvements though.

File: controllers/clustergroupupgrade_controller.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (imiller0) [reply controllers/clustergroupupgrade_controller.go:168]: That sounds great. I also agree that having all the policies showing compliant at the end of the batch is a good thing.

File: controllers/clustergroupupgrade_controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (imiller0) [reply controllers/clustergroupupgrade_controller.go:272]: Sure, it is fine to do that in a cleanup pass.

File: controllers/clustergroupupgrade_controller.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (imiller0) [reply controllers/clustergroupupgrade_controller.go:276]: Sounds good. Do you know how long it takes to step through each policy? Is each iteration on the order of seconds or minutes or ???

File: controllers/clustergroupupgrade_controller.go (hunk 5 above)
Comment (imiller0) [reply controllers/clustergroupupgrade_controller.go:1151]: I think we need to verify with the ACM team how/when the status gets updated. For example, when I first bind the cluster to the policy does it immediately get put into the NonCompliant status, or is there a phase where it is checking the cluster and the cluster is not found in either? Depending on their answer we may need to switch the test to be "positive compliance" rather than not-non-compliant.

File: controllers/clustergroupupgrade_controller.go (hunk 7 above)
Comment (imiller0) [reply controllers/clustergroupupgrade_controller.go:1185]: Getting things functional is the first (most critical) step right now. We should keep this on our cleanup list and make use of the libraries where we can (after things are working)

File: controllers/clustergroupupgrade_controller.go (hunk 1 above)
Comment (imiller0) [reply controllers/clustergroupupgrade_controller.go:598]: I haven't followed through the naming logic but we need to be sure that we handle various possible name-collisions:

- Multiple clusters with the same named policy
- Multiple upgrade CRs enforcing the same policy on different clusters
- Multiple upgrade CRs enforcing the same policy on the same cluster (likely an error case but it is possible to configure this and we should not break)

File: controllers/clustergroupupgrade_controller.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go]: We decide this by setting the nextReconcile variable. If we reconcile after an hour, then it's not good because we're basically waiting for an hour for a policy that we don't actually apply. I'm actually debating myself that it might be better to only include the next policy for which the policy is NonCompliant. This way we are not wasting reconciling iterations on policies for which the cluster is compliant... We do use a bit more time to get that info up front, but we save from the total upgrade time. Yep.

File: controllers/clustergroupupgrade_controller.go [hunk 18]
```diff
@@ -564,6 +702,7 @@ func (r *ClusterGroupUpgradeReconciler) ensureBatchPolicies(ctx context.Context,
 	return policies, nil
 }
 
+// TODO: remove this function as part of cleanup
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go]: Removed.

File: controllers/clustergroupupgrade_controller.go (hunk 5 above)
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:877]: Still waiting on answer from ACM. Meanwhile, I think it's safe to assume that if we can't access the status part of a policy that includes the clusters (for whatever reason), it's safer to assume the cluster is not compliant actually. That way, even if the cluster is already compliant with the policy, it won't impact the end result.

File: api/v1alpha1/clustergroupupgrade_types.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (irinamihai) [reply api/v1alpha1/clustergroupupgrade_types.go]: Done

File: config/crd/bases/ran.openshift.io_clustergroupupgrades.yaml [hunk 19]
```diff
@@ -44,7 +44,9 @@ spec:
               deleteObjectsOnCompletion:
                 type: boolean
               enable:
-                default: false
+                default: true
+                description: 'INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+                  Important: Run "make" to regenerate code after modifying this file'
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply config/crd/bases/ran.openshift.io_clustergroupupgrades.yaml]: Done

File: controllers/clustergroupupgrade_controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go]: Done

File: controllers/clustergroupupgrade_controller.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go]: Done

File: controllers/clustergroupupgrade_controller.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:307]: Done

File: controllers/clustergroupupgrade_controller.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:174]: Done - added method comment.

File: controllers/clustergroupupgrade_controller.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go]: Done

File: controllers/clustergroupupgrade_controller.go [hunk 20]
```diff
@@ -228,7 +212,7 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
 					Type:    "Ready",
 					Status:  metav1.ConditionTrue,
 					Reason:  "UpgradeCompleted",
-					Message: "The ClusterGroupUpgrade CR has all upgrade policies compliant",
+					Message: "The ClusterGroupUpgrade CR has all the upgrade policies compliant",
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go]: Done

File: controllers/clustergroupupgrade_controller.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go]: Decided to consider only the NonCompliant policies to save some upgrade time. :crossed_fingers:

File: controllers/clustergroupupgrade_controller.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go:451]: Done

File: controllers/clustergroupupgrade_controller.go [hunk 21]
```diff
@@ -318,55 +468,49 @@ func (r *ClusterGroupUpgradeReconciler) doManagedPoliciesExist(ctx context.Conte
...
 
-		// Update the list of new policy names.
-		newPolicies = append(newPolicies, newPolicy.GetName())
+	// Set new policy annotations - copy them from the managed policy.
+	newPolicy.SetAnnotations(managedPolicy.GetAnnotations())
+
+	// Set new policy remediationAction.
+	newPolicy.Object["spec"] = managedPolicy.Object["spec"]
+	specObject := newPolicy.Object["spec"].(map[string]interface{})
+	specObject["remediationAction"] = "enforce"
```
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go]: Done

File: controllers/clustergroupupgrade_controller.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go]: Done

File: controllers/clustergroupupgrade_controller.go (hunk 1 above)
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go]: I think we're good right now, but will keep an eye on this when we're doing more comprehensive tests.

File: controllers/clustergroupupgrade_controller.go (hunk 2 above)
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go]: Done

File: controllers/clustergroupupgrade_controller.go (hunk 3 above)
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go]: Done

File: controllers/clustergroupupgrade_controller.go (hunk 4 above)
Comment (irinamihai) [reply controllers/clustergroupupgrade_controller.go]: Done

**Review by nishant-parekh — APPROVED** [review]

File: controllers/clustergroupupgrade_controller.go (hunk 7 above)
Comment (Missxiaoguo) [reply controllers/clustergroupupgrade_controller.go:911]: > 
@irinamihai When I try to use the ManagedCluster from open-cluster-management API, the issue I have probably is same with what you are having here.

The error I have is, 
ERROR   setup   problem running manager {"error": "no kind is registered for the type v1.ManagedCluster in scheme \"pkg/runtime/scheme.go:100\""}

I figured it out that you need to register the scheme for the APIs for mapping bewteens Kinds and their corresponding Go types.

Basically, in this example to use multicloud-operators-placementrule apps APIs,
in the main.go,
// import the pkg
import (
     appsv1 github.com/open-cluster-management/multicloud-operators-placementrule/pkg/apis/apps/v1
) 
// register the scheme
func init() {
  ...
  utilruntime.Must(appsv1.AddToScheme(scheme))
}

Then, the call appsv1.xxx should work.
Hope this is helpful.
